plugins/discourse-presence/assets/javascripts/discourse/lib/presence-manager.js.es6

```diff
@@ -0,0 +1,197 @@
+          this._debounceClearUsers("users");
+          break;
+        case EDITING:
+          this._appendUser(this.get("editingUsers"), user, {
+            post_id: parseInt(message.post_id, 10)
+          });
+
+          this._debounceClearUsers("editingUsers");
+          break;
+        case CLOSED:
+          this._removeUser(user);
+          break;
+      }
+    });
+
+    this.set("subscribed", true);
+  },
+
+  unsubscribe() {
+    this.get("messageBus").unsubscribe(this.get("channel"));
+    this.set("subscribed", false);
+  },
+
+  @computed("topic.id")
+  channel(topicId) {
+    return `/presence/${topicId}`;
+  },
+
+  @observes("users.@each.last_seen")
+  updatePresenceUsers() {
+    this._debounceSetProperties({
+      presenceUsers: this.get("users")
+    });
+  },
+
+  @observes("editingUsers.@each.last_seen")
+  updateEditingPresenceUsers(users) {
+    this._debounceSetProperties({
+      editingPresenceUsers: this.get("editingUsers")
+    });
+  },
+
+  throttlePublish(state, whisper, postId) {
+    return throttle(
+      this,
+      this.publish,
+      state,
+      whisper,
+      postId,
+      KEEP_ALIVE_DURATION_SECONDS * 1000
+    );
+  },
+
+  publish(state, whisper, postId) {
+    const data = {
+      state,
+      topic_id: this.get("topic.id")
+    };
+
+    if (whisper) {
+      data.is_whisper = 1;
+    }
+
+    if (postId) {
+      data.post_id = postId;
+    }
+
+    return ajax("/presence/publish", {
+      type: "POST",
+      data
+    });
+  },
+
+  _debounceSetProperties(properties) {
+    debounce(this, this.setProperties, properties, 200);
+  },
+
+  _debounceClearUsers(users) {
+    return debounce(
+      this,
+      this.set,
+      users,
+      [],
+      (KEEP_ALIVE_DURATION_SECONDS + BUFFER_DURATION_SECONDS) * 1000
+    );
+  },
+
+  _removeUser(user) {
+    [this.get("users"), this.get("editingUsers")].forEach(users => {
+      const existingUser = users.findBy("id", user.id);
+      if (existingUser) users.removeObject(existingUser);
+    });
+  },
+
+  _appendUser(users, user, attrs) {
+    // Before appending users, we also want to find users that are stale and
+    // remove them.
+    const invalidUsers = [];
+    let existingUser;
+    let usersLength = 0;
+
+    users.forEach(u => {
+      if (u.id === user.id) {
+        existingUser = u;
+      } else if (
+        u.last_seen <=
+        Date.now() - KEEP_ALIVE_DURATION_SECONDS * 1000
+      ) {
+        invalidUsers.push(u);
+      }
+
+      if (attrs && attrs.post_id) {
+        if (u.post_id === attrs.post_id) usersLength++;
+      } else {
+        usersLength++;
+      }
+    });
+
+    users.removeObjects(invalidUsers);
+
+    const props = attrs || {};
+    props.last_seen = Date.now();
+
+    if (existingUser) {
+      existingUser.setProperties(props);
+    } else {
+      const limit = this.get("siteSettings.presence_max_users_shown");
+
+      if (usersLength < limit) {
```

Comment:
This acts a guard to ensure that memory doesn't blow up as we store the state of users in the `users` array. For `editingUsers`, we can technically still end up storing a lot of users in the array if everyone edits at the same time which seems kind of rare so I'm not tackling it yet. If we're worried about the memory growth here, we can't rely on the same messageBus channel that is used for replying to a topic.